Mathlib/CategoryTheory/Sites/RegularExtensive.lean

```diff
@@ -193,4 +200,159 @@ lemma extensive_regular_generate_coherent [Preregular C] [Extensive C] [Precoher
     | top => apply Coverage.saturate.top
     | transitive Y T => apply Coverage.saturate.transitive Y T<;> [assumption; assumption]
 
+section RegularSheaves
+
+variable {C}
+
+open Opposite
+
+/-- A presieve is *regular* if it consists of a single effective epimorphism. -/
+class Presieve.regular {X : C} (R : Presieve X) : Prop where
+  /-- `R` consists of a single epimorphism. -/
+  single_epi : ∃ (Y : C) (f : Y ⟶ X), R = Presieve.ofArrows (fun (_ : Unit) ↦ Y)
+    (fun (_ : Unit) ↦ f) ∧ EffectiveEpi f
+
+/--
+The map to the explicit equalizer used in the sheaf condition.
+-/
+def MapToEqualizer (P : Cᵒᵖ ⥤ Type (max u v)) {W X B : C} (f : X ⟶ B)
+    (g₁ g₂ : W ⟶ X) (w : g₁ ≫ f = g₂ ≫ f) :
+    P.obj (op B) → { x : P.obj (op X) | P.map g₁.op x = P.map g₂.op x } :=
+  fun t ↦ ⟨P.map f.op t, by
+    change (P.map _ ≫ P.map _) _ = (P.map _ ≫ P.map _) _;
+    simp_rw [← P.map_comp, ← op_comp, w] ⟩
+
+/--
+The sheaf condition with respect to regular presieves, given the existence of the relavant pullback.
+-/
+def EqualizerCondition (P : Cᵒᵖ ⥤ Type (max u v)) : Prop :=
+  ∀ (X B : C) (π : X ⟶ B) [EffectiveEpi π] [HasPullback π π], Function.Bijective
+    (MapToEqualizer P π (pullback.fst (f := π) (g := π)) (pullback.snd (f := π) (g := π))
+    pullback.condition)
+
+/--
+The `FirstObj` in the sheaf condition diagram is isomorphic to `F` applied to `X`.
+-/
+noncomputable
+def EqualizerFirstObjIso (F : Cᵒᵖ ⥤ Type (max u v)) {B X : C} (π : X ⟶ B) :
+    Equalizer.FirstObj F (Presieve.singleton π) ≅ F.obj (op X) :=
+  CategoryTheory.Equalizer.firstObjEqFamily F (Presieve.singleton π) ≪≫
+  { hom := fun e ↦ e π (Presieve.singleton_self π)
+    inv := fun e _ _ h ↦ by
+      induction h with
+      | mk => exact e
+    hom_inv_id := by
+      funext _ _ _ h
+      induction h with
+      | mk => rfl
+    inv_hom_id := by aesop }
+
+instance {B X : C} (π : X ⟶ B) [EffectiveEpi π] [HasPullback π π] :
+    (Presieve.singleton π).hasPullbacks where
+  has_pullbacks hf _ hg := by
+    cases hf
+    cases hg
+    infer_instance
+
+/--
+The `SecondObj` in the sheaf condition diagram is isomorphic to `F` applied to the pullback of `π` 
+with itself
+-/
+noncomputable
+def EqualizerSecondObjIso (F : Cᵒᵖ ⥤ Type (max u v)) {B X : C} (π : X ⟶ B) [EffectiveEpi π]
+    [HasPullback π π] :
+    Equalizer.Presieve.SecondObj F (Presieve.singleton π) ≅ F.obj (op (Limits.pullback π π)) :=
+  Types.productIso.{max u v, max u v} _ ≪≫
+  { hom := fun e ↦ e (⟨X, ⟨π, Presieve.singleton_self π⟩⟩, ⟨X, ⟨π, Presieve.singleton_self π⟩⟩)
+    inv := fun x ⟨⟨_, ⟨_, h₁⟩⟩ , ⟨_, ⟨_, h₂⟩⟩⟩ ↦ by
+      induction h₁
+      induction h₂
+      exact x
+    hom_inv_id := by
+      funext _ ⟨⟨_, ⟨_, h₁⟩⟩ , ⟨_, ⟨_, h₂⟩⟩⟩
+      induction h₁
+      induction h₂
+      rfl
+    inv_hom_id := by aesop }
+
+lemma isSheafFor_regular_of_hasPullbacks {B : C} {S : Presieve B} [S.regular] [S.hasPullbacks]
+    {F : Cᵒᵖ ⥤ Type (max u v)}
+    (hFecs : EqualizerCondition F) : S.IsSheafFor F := by
```

Comment:
I think this can be renamed to enable dot-notation
```suggestion
lemma EqualizerCondition.isSheafFor {B : C} {S : Presieve B} [S.regular] [S.hasPullbacks]
    {F : Cᵒᵖ ⥤ Type (max u v)}
    (hFecs : EqualizerCondition F) : S.IsSheafFor F := by
```
Or is that a bad idea?

What about the converse? Is this an if-and-only-if?